### ISSUE TITLE
Dusty Cue pays the emulator on every gradient: tabulate the nebular continuum as a spectrum so the grid can serve a model with dust

### ISSUE THREAD

**Issue by suchethac** [suchethac/tengri#2387]
On a dusty Cue model the nebular grid is disarmed entirely, so every likelihood evaluation calls the Cue emulator. **The emulator is ~97% of the cost of such a fit.** The grid could serve these models if it tabulated the nebular *continuum* as a spectrum instead of only band-integrated photometry — the measurements below say what that would take.

## Why it is disarmed today

`_nebular_continuum_consumers` states the rule:

> Serving photometry from the per-Q_H grid requires zeroing `sed_nebular`, so it is available only when nothing downstream reads the continuum. A non-empty result sets `must_materialize_sed` and disarms the shortcut.

`NebularGridTable` carries `log_line_per_qh` (line luminosities) and `log_phot_per_qh` / `log_restband_per_qh` — the nebular contribution **already integrated through each filter**. A band-integrated flux cannot be attenuated after the fact, because attenuation does not commute with band integration:

∫F(λ)e^(−τ(λ))R(λ)dλ ≠ (∫F(λ)R(λ)dλ)·e^(−τ(λ_eff))

Pre-#1281 tengri effectively did that, measured at **0.41%** photometry error against a **0.0115%** dust-free floor. #1671 is why that matters: a constant forward bias enters the posterior gradient multiplied by SNR, so better data makes it worse. The disarm is correct; it is the *granularity* that is wrong.

There is no way to sidestep it by fitting lines alone: `data_type` is only `photometry` / `spectroscopy` / `joint`, so a line-flux channel always rides on a continuum channel that needs the materialized SED.

## What it costs

Gradient FLOPs of the fit objective, same dusty model, `approx="auto"`, only the nebular backend changed:

| `neb` | gradient FLOPs |
| --- | --- |
| `{'type': 'cue'}` | **50,788,420** |
| `{'type': 'none'}` (baked-in) | **1,308,764** |

End to end, `mcmc_nuts_fast` on notebook 10's model (Cue + dust + 10 measured lines, D=9 after `profile_mass`): **210.8 s** cold. The same model **without dust**, where the grid does engage: **37.5 s**, of which the grid itself is 2.9x (109.3 s → 37.5 s).

## Proposed fix

Invert `must_materialize_sed` from *"disarm the grid"* to *"use the grid to materialize"*. Add a continuum channel to `NebularGridTable`, reconstruct `sed_nebular` as a spectrum (tabulated continuum + the already-tabulated lines), and let dust attenuate it normally. The saving preserved is skipping the emulator, which is the 97%; materializing the continuum is something the dusty path already does today.

`CueBackend.predict_nebular_continuum` / `predict_continuum_jit` already exist, so the continuum is separately available to tabulate.

### Measured design inputs

**1. The continuum shape varies over the gridded axes, so it must be gridded — not held fixed.** Shape deviation from the prior-median continuum, lines masked at ±250 km/s, normalized to unit integral:

| axis | max shape deviation | continuum amplitude |
| --- | --- | --- |
| `met_logzsol` | 8.50% | 0.64–1.50x |
| `neb_logU` | 17.73% | **0.993–1.008x** |
| `neb_logZ_gas` | 64.32% | 0.84–1.30x |
| all three, 80 corners | **109.56%** | — |

These are exactly the axes `precompute_nebular_grid` already spans, so this is the same interpolation the lines get. Note `neb_logU` moves the shape by 18% while leaving the amplitude flat to <1% — an amplitude-only approximation would be silently wrong.

**2. Nodes must be placed at the recombination edges, not log-spaced.** At 200 log-spaced nodes the median error is 0.006% but the **max is 72.9%**, and it is not scattered:

| wave [Å] | rel err | nearest edge |
| --- | --- | --- |
| 3640.0 | 72.9% | Balmer 3646 (−0.2%) |
| 3639.6 | 72.6% | Balmer 3646 (−0.2%) |
| 3640.5 | 72.3% | Balmer 3646 (−0.1%) |

All 20 worst pixels lie within 0.4% of the **Balmer jump**. 94.2% of the summed error sits within 5% of a recombination edge, on 246 of 3413 pixels. The continuum is *piecewise* smooth; a node straddling a jump interpolates across a discontinuity. Max error away from the edges I tested is still 26.5%, so the node set needs the full hydrogen series limits (912, 3646, 8204, 14588, 22794 Å, …) plus the He I / He II edges, with log spacing between them.

**3. Storage is not a constraint once nodes are placed well** (float32):

| nodes | 8³ grid | 16³ grid |
| --- | --- | --- |
| 100 | 0.2 MB | 1.6 MB |
| 200 | 0.4 MB | 3.3 MB |
| 400 | 0.8 MB | 6.6 MB |
| full 6185 | 13 MB | 101 MB |

## Acceptance gate

This is the seam that produced #1281, #1665, #1671, #1748, #1770, #1943, #2222 and #2235, so the bar is a measurement, not a green test run:

- Dusty photometry from the reconstructed continuum must match the exact path to **better than the 0.0115% dust-free floor** — the number #1281 rejected the old shortcut against was 0.41%.
- Error must be reported **per band**, not as an average: the Balmer jump sits inside `des_g`, so a mean over bands hides exactly the failure mode this design has.
- A guard that the node set actually contains the recombination edges, so a later refactor cannot silently revert to log spacing.
- Gradient FLOPs before/after on a dusty Cue model, to confirm the emulator really left the graph (equal FLOPs = the config never reached it — this repo's own signature).

## Scope

`NebularGridTable`, `precompute_nebular_grid`, the `use_grid` / `must_materialize_sed` branch in `SEDModel`, and `fast_nebular_can_engage` + the advice that quotes it. Applies to every Q_H-linear backend, not only Cue (CloudyGrid shares the path).

**Comment by suchethac** [comment]
Design constraint found while coordinating with the #2346 lane (Cue continuum red tail), worth pinning before implementation.

**The fast path publishes `sed_nebular` as identically zero.** `components/nebular/component.py` ~line 712:

```python
use_grid = (self.grid_table is not None and ... and not self.must_materialize_sed)
if use_grid:
    nebular_sed = zeros
```

So today, `approx=FeaturePrecomp()` on a dust-free Cue model makes the component publish a zero continuum array while the backend is very much a continuum publisher — the continuum is served from the per-Q_H table instead. Any downstream rule that asks "does this model have a nebular continuum?" by **inspecting the array** therefore answers differently for the same physical model depending only on whether an approximation was requested. That is an approximation flag silently changing physics, the #1748/#1770 failure class. Such rules must key on the backend's declared capability, not on the published array and not on `must_materialize_sed`.

This gets sharper after this issue lands, not softer: the entire point here is that a dusty model reconstructs a **real** continuum from the table, so the array-inspection answer flips a second time. Concretely it affects the #2346 auto-rule (`radio include_freefree=None` resolving against "the nebular backend publishes a continuum"), which I have flagged to that lane.

**Second constraint, same coordination:** the continuum channel must be built through `model.predict_state`, which is how `precompute_nebular_grid` already builds every node (~line 890), rather than through `CueBackend.predict_nebular_continuum` (which returns on Cue's native `cont_wav`). Using the native accessor would create a second place where the continuum shape is decided, so a change at the `jnp.interp(ssp_wave, cont_wav, cont_lum, left=0.0, right=0.0)` site in `cue.py` — exactly what #2346 is changing — would leave the table describing a different SED from the exact path. Going through `predict_state` means the table inherits such changes automatically.

**Third, a scoping note on the node budget above.** The 200-node / 0.006% median / 72.9% max figures were measured with the comparison mask restricted to 1000-40000 Å. They do not cover a continuum extended toward ~1 m. Tabulating `log L_nu` against `log lambda` keeps that cheap — an optically thin free-free tail is a power law and so is exactly linear in log-log — but the node set must be stated in those terms rather than inheriting a fixed count.

**Comment by suchethac** [comment]
**The ~97% emulator share is a property of Cue, not of one fixture — re-measured on three SSP grids.**

I quoted 50,788,420 (cue) against 1,308,764 (baked-in) on a single grid and called it ~97%. That deserved a check, because the #2346 lane measured the same nominal fit at a 6.5x different absolute baseline purely from the SSP grid. Same model on each grid (dusty two-component Calzetti, 4 bands, photometry only, `approx=WavePrecomp()` at fit time, only `neb` changing between arms), gradient FLOPs of the fit objective:

| SSP grid | n_wave | `neb='cue'` | `neb='none'` | emulator share |
| --- | --- | --- | --- | --- |
| `fsps_prsc_miles_chabrier` | 5994 | 50,788,420 | 1,308,764 | **97.4%** |
| `bc03_pdva_stelib_chabrier` | 6900 | 61,346,520 | 2,462,270 | **96.0%** |
| `fsps_mist_c3k_a_chabrier` | 11149 | 93,725,352 | 1,440,861 | **98.5%** |

So the headline figure holds, and the payoff argument for this issue stands.

**My reason for expecting otherwise was wrong, and the way it is wrong is worth recording.** I predicted the share would shrink on finer grids, since the emulator's cost tracks the ionization axes rather than the wavelength axis. It does not — the finest grid gives the *highest* share. The emulator path is itself O(n_wave): its continuum is interpolated onto `ssp_wave` on every evaluation (`jnp.interp(ssp_wave, cont_wav, cont_lum, ...)` at `cue.py`), so numerator and denominator grow together.

That has a direct consequence for the design in this issue, which I had not drawn: **the continuum table must be interpolated onto the model grid at lookup time too**, so the reconstructed path keeps an O(n_wave) term of its own. The saving is the emulator *evaluation*, not the wavelength interpolation, and the acceptance measurement must therefore be a before/after on the same grid rather than a comparison against the baked-in floor — the baked-in numbers above (1.3-2.5M) are **not** the target this design can reach.

All figures in the issue body should be read as grid-attributed; the original 50.8M/1.3M pair is the 5994-node PRSC grid.

**Comment by suchethac** [comment]
**The acceptance question is UNRESOLVED, and I want that on the record before anyone implements this.** I attempted the per-band accuracy measurement twice and both harnesses are unsound. Neither result should be used.

**Attempt 1 — line-masking artifact.** I built the continuum by masking 138 catalog lines out of `sed_nebular` (±5×250 km/s) and filling the gaps by log-log interpolation. Result: worst band error 0.09% (log-spaced, 400 nodes) and 0.18% (edge-aware, 420). Suspect because the fill introduces a kink at every mask boundary, so the reconstruction error may be of my own artifact — and the tell is that the worst band was `wise_w1`/`des_z`, **away** from the Balmer jump, the opposite of what the physics predicts.

**Attempt 2 — non-monotonic, therefore buggy.** Using `CueBackend.predict_nebular_continuum` (the true line-free continuum, 1841 native points, 915 Å–1e8 Å), worst band error by node count: 93.4% (100) → 31.5% (200) → 4.85% (400) → **12.9% (800)**. Error that *rises* when nodes are added is a harness bug, full stop. Two candidates: I hand-rolled `log_z = met_logzsol + LOG10_ZSUN` and omitted the `k` factor (`lyc_dust_escape_factor(neb_fesc, neb_fdust)`) that `cue.py` applies to `cont_lum`, so my "true continuum" is probably not the one the model uses; and I integrate narrow filter curves on a grid spanning five decades, where the band integral is coarsely sampled.

**What is NOT in doubt** (measured, stable, reproduced):

- The emulator is ~97% of a dusty fit, across three SSP grids (97.4% / 96.0% / 98.5%). The payoff is real.
- The continuum shape varies up to 110% across `_CANDIDATE_AXES`, so it must be gridded on them, not held fixed.
- The continuum has recombination-edge discontinuities, and at 200 log-spaced nodes all 20 worst *pixels* lie within 0.4% of the Balmer jump, with 94.2% of summed error within 5% of an edge. That part was unambiguous.

**What is in doubt:** whether *any* node count meets a per-band bar under the 0.0115% dust-free floor, and whether edge-aware placement helps at the band level. In attempt 1 it helped slightly at low node counts and *hurt* at 400 — which, if real, would mean the pixel-level edge story does not translate to band-integrated flux, because a filter integrates across the jump either way.

**Before implementation, someone must build a harness that passes its own sanity checks:** error monotonically decreasing in node count, the continuum obtained the way the model obtains it (through `predict_state`, not a hand-assembled backend call), and filters integrated on a grid fine enough that the quadrature is not the error source. Until that exists, the acceptance gate in the issue body is aspirational rather than demonstrated, and this issue should not be picked up as ready-to-build.

**Comment by suchethac** [comment]
#2419 landed (main `8e642ab11`, closing #2346) and it **changes the function this issue proposes to tabulate**, so the node-budget numbers recorded here earlier are superseded. Noting that before anyone — including a later me — builds on them.

**What moved.** `NEBULAR_CONTINUUM_WAVE_MAX` is now `1e10` Å (`utils/wavelength.py:29`): the Cue/CloudyGrid continuum continues past 1 cm instead of ending where the tables do, via `interp_continuum_with_freefree_tail` (`components/nebular/_shared.py:607`), with `nebular_backend_carries_freefree` (`components/nebular/_models.py:70`) as the auto-rule predicate. Both the Cue and CloudyGrid call sites now share that single continuum-shape source.

**Why the earlier numbers no longer apply.** The node sweep quoted on this issue (200 log-spaced nodes: median band error 0.006%, max 72.9%, all 20 worst pixels within 0.4% of the Balmer jump, 94.2% of error within 5% of a recombination edge) was measured against a continuum ending around 1e8 Å. The axis is now up to two decades longer. At a fixed node count, log-spaced nodes per decade fall by roughly a quarter going from six decades to eight — so "200 nodes" is not the same tabulation it was, and the error figures have to be re-measured rather than rescaled.

The qualitative finding should survive: a free-free tail is a smooth power law and is cheap to represent, so the error should still concentrate at the hydrogen/helium recombination edges rather than spread along the new range. But that is a prediction, not a measurement, and this issue has already been bitten twice by harnesses that looked reasonable.

**The accuracy gate remains UNRESOLVED**, as recorded earlier. Both harnesses built for it were unsound — one had a line-masking-and-fill artifact whose worst band moved *away* from the Balmer jump (the opposite of what the physics predicts), the other gave non-monotonic error with node count (4.85% at n=400 rising to 12.9% at n=800), which is a harness bug rather than a property of the continuum. Neither result was reported as a finding and neither should be resurrected.

**What would unblock it.** The dust/IGM lane's #2439 work is building a per-band edge-error harness against an analytic-break template with a closed-form band integral, worst-band with the straddle asserted. That is the right shape of instrument for this question and it is better to adopt it than to build a third unsound one here. Sequencing this issue behind that, and behind a re-measurement on the post-#2419 continuum.
